### PR TITLE
dnsdist: Implement Channels for communication between threads

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Internally this uses the same mechanism as before, passing pointers over a pipe, but the new classes make that easier to use:
- clear separation between sender and receiver
- clear ownership of the descriptor
- less code duplication
- 
### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
Interestingly TSAN now thinks there is a race in the dnsdist code handling TC=1 responses received over UDP, for incoming queries received over DoH:
- the query is received in the DoH thread, sent to the backend over UDP (TSAN READ)
- the UDP responder thread picks up the response, notices it's truncated in `handleUDPResponseForDoH`, passes it to the DoH thread, which passes it to a TCP worker thread
- the worker thread prepend the size of the query in the query buffer, causing TSAN to detect a race (TSAN WRITE)

I don't think there is a race, unless I'm missing something, since the query has to have been sent by the time the corresponding response is received. It might take some annotation magic to convince TSAN, though.

**Comment by rgacogne** [comment]
Sounds a lot like https://github.com/google/sanitizers/issues/1498
And indeed the actual range check is done after `send()` has returned: https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc#L6982

**Comment by rgacogne** [comment]
Disabling the `send` wrappers via `intercept_send=0` works around that issue. This is a bit sad but better than the alternatives, I guess.

**Comment by rgacogne** [comment]
Marking this as "ready for review" to get feedback.

File: pdns/channel.hh
Comment (omoerbeek) [inline pdns/channel.hh:271]:
```suggestion
      T* objPtr{nullptr};
```

File: pdns/channel.hh
Comment (omoerbeek) [inline pdns/channel.hh:295]: Does it make sense to test for write returning != writesize on the write side as well? It *should* never happen, but still...

**Review by omoerbeek — COMMENTED** [review]
Two small comments, will do a full review soon.

File: pdns/channel.hh
Comment (rgacogne) [reply pdns/channel.hh:295]: I think we do, don't we?

Comment (omoerbeek) [reply pdns/channel.hh:295]: Ah right, just the zero case is not handled as a separate case. errno is not set if write returns zero, so to be pedantic the zero case should be handled separately,.

Comment (rgacogne) [reply pdns/channel.hh:295]: Agreed, I'll do that!

File: pdns/dnsdistdist/dnsdist-tcp-upstream.hh
Comment (omoerbeek) [inline pdns/dnsdistdist/dnsdist-tcp-upstream.hh:6]:
```suggestion
struct TCPCrossProtocolResponse;
```

File: pdns/channel.hh
Comment (omoerbeek) [inline pdns/channel.hh:263]:
```suggestion
    return receive(D());
```

**Review by omoerbeek — CHANGES_REQUESTED** [review]

**Comment by rgacogne** [comment]
Comments applied, thanks!

File: pdns/channel.hh
Comment (omoerbeek) [inline pdns/channel.hh:141]:
```suggestion
  std::pair<Sender<T, D>, Receiver<T, D>> createObjectQueue(bool sendNonBlocking = true, bool receiveNonBlocking = true, size_t pipeBufferSize = 0, bool throwOnEOF = true);
```

**Review by omoerbeek — APPROVED** [review]
Looks really nice, much more easier on the eye than the current code. Looking forward to using it in the recursor a well.

**Comment by rgacogne** [comment]
Added unit tests.

**Comment by rgacogne** [comment]
[d5eaeb3](https://github.com/PowerDNS/pdns/pull/11427/commits/d5eaeb3337d0e98db6e1548a4a9ab607082c15fd) broke the auth build, so I will have to squash the corresponding fixup commit [e8dca71](https://github.com/PowerDNS/pdns/pull/11427/commits/e8dca71d79009aaf972264b07ef290547aa8b850) before merging this PR.

**Comment by rgacogne** [comment]
Rebased to fix conflicts.

**Comment by rgacogne** [comment]
Otto, Fred, how would you feel about merging this PR?

**Comment by fredmorcos** [comment]
> Otto, Fred, how would you feel about merging this PR?

It's been on my todo for a while. I need to go through it still. I'll do that first thing on Monday.

**Review by fredmorcos — APPROVED** [review]
Just a few nitpicks, nothing important. Generally I really like this and I guess you know why :-D. It did clean up quite a bit.

Also please wait for a review from @omoerbeek as well.

File: pdns/doh.hh
Comment (rgacogne) [reply pdns/doh.hh:253]: Unfortunately it wouldn't, because we have a non-static `release()` member function and a static `release(DOHUnit*)` one, so the compiler cannot decide which one we are referring to.

File: pdns/channel.cc
Comment (rgacogne) [reply pdns/channel.cc:25]: That's weird because the CI's `clang-tidy` actually complained about the initial `namespace pdns { namespace channel`, leading me to switch to the new form in https://github.com/PowerDNS/pdns/pull/11427/commits/35b27ac8d18b43702a63e4078b68776c66f561ab : https://clang.llvm.org/extra/clang-tidy/checks/modernize/concat-nested-namespaces.html

**Review by omoerbeek — CHANGES_REQUESTED** [review]
This is really nice! 
A thing: clang-tidy does report some issues (as shown in the build logs), but these are not detected in the summary task. Wonder why....

Some of these issues might indicate a real problem like `d_fd` not being initialised in all cases, reported in https://github.com/PowerDNS/pdns/actions/runs/5253429184/jobs/9490749878 dnsdist build log.

File: pdns/channel.cc
Comment (fredmorcos) [reply pdns/channel.cc:25]: Ah. I think I figured out why, creating a git worktree with magit - which I usually do to review PRs - doesn't copy over ignored files. So my clang-tidy was complaining using the default set of configuration, in this particular case saying that nested namespace declarations are a C++17 extension.

File: pdns/channel.cc
Comment (fredmorcos) [inline pdns/channel.cc:28]: clang-tidy should complain about the identifier being too short here...

**Comment by rgacogne** [comment]
> A thing: clang-tidy does report some issues (as shown in the build logs), but these are not detected in the summary task. Wonder why....
> 
> Some of these issues might indicate a real problem like `d_fd` not being initialised in all cases, reported in https://github.com/PowerDNS/pdns/actions/runs/5253429184/jobs/9490749878 dnsdist build log.

I'll have a second look tomorrow but at the moment I cannot make sense of this warning, as far as I can tell `d_fd` cannot be uninitialized.

File: pdns/channel.cc
Comment (rgacogne) [reply pdns/channel.cc]: Fixed!

**Comment by rgacogne** [comment]
> > A thing: clang-tidy does report some issues (as shown in the build logs), but these are not detected in the summary task. Wonder why....
> > Some of these issues might indicate a real problem like `d_fd` not being initialised in all cases, reported in https://github.com/PowerDNS/pdns/actions/runs/5253429184/jobs/9490749878 dnsdist build log.
> 
> I'll have a second look tomorrow but at the moment I cannot make sense of this warning, as far as I can tell `d_fd` cannot be uninitialized.

I still don't see it :-/ Otto, Fred, did you look at the reported issue or just at the report?

**Comment by omoerbeek** [comment]
> I still don't see it :-/ Otto, Fred, did you look at the reported issue or just at the report

I did not look at the details of the report, doing that now.

**Comment by omoerbeek** [comment]
> I did not look at the details of the report, doing that now.

I cannot spot the issue either. It seems to think that the default ct of `Receiver` does not init `d_fd`, but `d_fd` is of type `FDWrapper`, which itself has a default  value for it's only field `d_fd` of type `int`.

**Review by omoerbeek — APPROVED** [review]

**Pull request merged**
